core/src/main/java/feign/SharedParameters.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2012-2021 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign;
+
+import feign.InvocationHandlerFactory.MethodHandler;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+public class SharedParameters {
+
+  private final Map<String, Object> paramNameToValue = new HashMap<>();
+
+  public ConfigurationMethodHandler newHandler(final MethodMetadata methodMetadata) {
+    return new ConfigurationMethodHandler(methodMetadata);
+  }
+
+  public Map<String, Object> asMap() {
+    return paramNameToValue;
+  }
+
+  public class ConfigurationMethodHandler implements MethodHandler {
+
+    private final MethodMetadata methodMetadata;
+
+    public ConfigurationMethodHandler(final MethodMetadata methodMetadata) {
+      this.methodMetadata = methodMetadata;
+    }
+
+    @Override
+    public Object invoke(Object[] argv) throws Throwable {
+      IntStream.range(0, argv.length)
+          .forEach(index -> methodMetadata.indexToName()
+              .get(index)
+              .forEach(name -> paramNameToValue.put(name, argv[index])));
+      return null;
+    }
```

Comment:
This could overwrite a value placed on the method itself.  We'll need to consider what to do if a user defined a shared value explicitly in a method.  What should occur here?  Should the shared value always overwrite or would we want to consider the one explicitly declared?

Reply:
In the example below, I reckon the intent of the user would be to have the `env` parameter passed in `getResourceWithExplicitEnv` take precedence over any value set by `setEnv`. Thoughts?

```java
interface ClientApi {

  @Shared
  void setEnv(@Param("env") String env);

  @RequestLine("GET /{resource}/{env}")
  void getResource(@Param("resource") String resource);

  @RequestLine("GET /{resource}/{env}")
  void getResourceWithExplicitEnv(@Param("resource") String resource, @Param("env") String env);
}
```

This example deserves a unit test.